src/app/util/attribute-storage.cpp

```diff
@@ -646,7 +747,12 @@ Status emAfReadOrWriteAttribute(const EmberAfAttributeSearchRecord * attRecord,
                                 }
 
                                 // Internal storage is only supported for fixed endpoints
-                                if (!isDynamicEndpoint)
+                                // and dynamic ones with dynamicAttributeStorage assigned.
+                                if (!isDynamicEndpoint
+#if CHIP_DEVICE_CONFIG_DYNAMIC_ENDPOINT_COUNT > 0
```

Comment:
how about having `hasDynamicAttributeStorage` always exist and have a single ifdef (and if the count is 0, then it will always be `false`? ifdefs inside if clauses are hard to maintain.

Reply:
My goal was not to cause any code or storage size increase for the (majority) of non-dynamic use cases of this code.

We can of course drop that goal, and remove all the CHIP_DEVICE_CONFIG_DYNAMIC_ENDPOINT_COUNT ifdefs. But that would mean even static-only builds would have the dynamicAttributeStorage field (now a span, 2*size_t) so a RAM increase of 2*size_t*MAX_ENDPOINT_COUNT.

Reply:
I am asking for:

```cpp
#if CHIP_DEVICE_CONFIG_DYNAMIC_ENDPOINT_COUNT > 0
            bool hasDynamicAttributeStorage = !emAfEndpoints[ep].dynamicAttributeStorage.empty();
            if (hasDynamicAttributeStorage)
            {
                // endpoint storage is not in the static global `attributeData`, but offset
                // from the per-endpoint dynamicAttributeStorage pointer.
                // Endpoint processing starts here, so reset the offset.
                attributeStorageOffset = 0;
            }
#else
            const bool hasDynamicAttributeStorage = false;
#endif
```

Then you will not have a ifdef in the middle of an if and compiler should be able to optimize it out. I hope you will not get compiler warnings of unused or always false....

Reply:
It works with `check` on my mac, but it looks fragile to me regarding build environments it should work on - it depends on a compiler optimizing the dead code (lines 725..727) out before it tries to resolve the `dynamicAttributeStorage` field (which would not exist when `hasDynamicAttributeStorage` is const false).

I tried to look up the details, there's no such discard guarantee with a `const bool`, but in >=C++17 it should work with the bool declared as constexpr:

```C++
#if CHIP_DEVICE_CONFIG_DYNAMIC_ENDPOINT_COUNT > 0
            bool hasDynamicAttributeStorage = ...
#else
            constexpr bool hasDynamicAttributeStorage = false;
#endif
```

Should I settle for that?

Reply:
sure, that works. For me I am looking to avoid a split ifdef inside if brackets, since that seems harder to read.

Reply:
Unfortunately, while this works on some platforms, it fails exactly as I feared it could on others, that's why most of the examples have failed to build now:

```
INFO    ../../../../../src/app/util/attribute-storage.cpp: In function ‘chip::Protocols::InteractionModel::Status emAfReadOrWriteAttribute(const EmberAfAttributeSearchRecord*, const EmberAfAttributeMetadata**, uint8_t*, uint16_t, bool)’:
INFO    ../../../../../src/app/util/attribute-storage.cpp:741:75: error: ‘struct EmberAfDefinedEndpoint’ has no member named ‘dynamicAttributeStorage’
INFO      741 |                                     attributeLocation = emAfEndpoints[ep].dynamicAttributeStorage.data();
INFO          |                                                                           ^~~~~~~~~~~~~~~~~~~~~~~
```

…despite constexpr. Either that guarantee of not trying to compile code that is defined dead by a constexpr does not exist in C++ or some of the embedded platforms don't have C++17 yet.  So I'll revert that part back to the uglier (I agree with you) but less assuming #ifdefs.

Reply:
I am explicitly only asking for not having a ifdef in the middle of an if condition. Not asking for the other bits to be converted to a `if constexpr` part instead of ifdefs. You can keep those like you had them.